File: problems/14/solution.py
class Solution:
    def partitionLinkedList(self, head, N):
        if not head:
            return None

        p = ListNode(0)
        p.next = head
        head = p
        last = head
        if head.next == None:
            return head.next

        n = 0
        while last.next != None:
            last = last.next
            n = n + 1


        while n > 0:
            if p.next.val < N:
                p = p.next
                n = n - 1
            else:
                last.next = ListNode(p.next.val)
                last = last.next
                p.next = p.next.next
                n = n - 1

        return head.next

File: problems/14/solution.c
ListNode* partitionLinkedList(ListNode* head, int N) {
    if (!head) {
        return NULL;
    }

    ListNode *p = (ListNode*) malloc(sizeof(ListNode));
    p->val = 0;
    p->next = head;
    head = p;
    ListNode *last=head;

    if (head->next == NULL) {
        return head->next;
    }

    int n = 0;
    while (last->next != NULL) {
        last=last->next; n++;
    }

    while (n > 0) {
        if (p->next->val < N) {
            p=p->next;
            n--;
        } else {
            ListNode* tmpNode = (ListNode*) malloc(sizeof(ListNode));
            tmpNode->val = p->next->val;
            tmpNode->next = NULL;

            last->next = tmpNode;
            last = last->next;
            p->next = p->next->next;
            n--;
        }
    }
    return head->next;
}
